src/utils/communication_buffer.hpp

```diff
@@ -0,0 +1,418 @@
+  }
+}
+
+template <class T>
+void CommBuffer<T>::TryStartReceive() noexcept {
+#ifdef MPI_PARALLEL
+  if (*comm_type_ == BuffCommType::receiver && !*started_irecv_) {
+    PARTHENON_REQUIRE(
+        *my_request_ == MPI_REQUEST_NULL,
+        "Cannot have another pending request in a buffer that is starting to receive.");
+    if (!IsActive())
+      Allocate(); // For early start of Irecv, always need storage space even if not used
+    PARTHENON_MPI_CHECK(MPI_Irecv(buf_.data(), buf_.size(),
+                                  MPITypeMap<buf_base_t>::type(), send_rank_, tag_, comm_,
+                                  my_request_.get()));
+    *started_irecv_ = true;
+  } else if (*comm_type_ == BuffCommType::sparse_receiver && !*started_irecv_) {
+    int test;
+    MPI_Status status;
+    // This is the extra MPI call that impacts performance mentioned in Athena++
+    // PARTHENON_MPI_CHECK(MPI_Iprobe(MPI_ANY_SOURCE, MPI_ANY_TAG, MPI_COMM_WORLD, &test,
+    //                               MPI_STATUS_IGNORE));
+    // Check if our message is available so that we can use the correct buffer size
+    PARTHENON_MPI_CHECK(MPI_Iprobe(send_rank_, tag_, comm_, &test, &status));
+    if (test) {
+      // For optimal buffer memory useage, we can only post the Irecv once we know the
+      // size of the incoming buffer
+      int size;
+      PARTHENON_MPI_CHECK(MPI_Get_count(&status, MPITypeMap<buf_base_t>::type(), &size));
+      if (size > 0) {
+        if (!active_) Allocate();
+        PARTHENON_MPI_CHECK(MPI_Irecv(buf_.data(), buf_.size(),
+                                      MPITypeMap<buf_base_t>::type(), send_rank_, tag_,
+                                      comm_, my_request_.get()));
+      } else {
+        if (active_) Free();
+        PARTHENON_MPI_CHECK(MPI_Irecv(&null_buf_, 0, MPITypeMap<buf_base_t>::type(),
+                                      send_rank_, tag_, comm_, my_request_.get()));
+      }
+      *started_irecv_ = true;
+    }
+  }
+#endif
+}
+
+template <class T>
+bool CommBuffer<T>::TryReceive() noexcept {
+  if (*state_ == BufferState::received || *state_ == BufferState::received_null)
+    return true;
+
+  if (*comm_type_ == BuffCommType::receiver ||
+      *comm_type_ == BuffCommType::sparse_receiver) {
+#ifdef MPI_PARALLEL
+    (*nrecv_tries_)++;
+    if (*nrecv_tries_ > 1e6)
+      PARTHENON_FAIL("MPI probably hanging after 1e6 receive tries.");
+
+    TryStartReceive();
+
+    if (*started_irecv_) {
+      MPI_Status status;
+      int flag;
+      // This is the extra MPI call that impacts performance as mentioned in Athena++.
+      // Depending on the MPI implementation and the order of MPI_Test calls, as well
+      // as the total number of buffers being communicated, I have found this can have
+      // anywhere from no impact on the walltime to a factor of a few reduction in the
+      // walltime. It seems to be unpredictable as far as I can tell.
+      for (int i = 0; i < 1; ++i)
```

Comment:
This for "loop" can probably go?

Reply:
I wanted to leave it in there to remind people that you may want to call `MPI_Iprobe` more than once.